### PR TITLE
system-linux: add support to display 2.5G+ in network.device status

### PR DESCRIPTION
**Pull request by rthakur33**
should be merged post https://github.com/openwrt/netifd/pull/2

### CODE REVIEW

File: system-linux.c [hunk 1]
```diff
@@ -1753,6 +1753,10 @@ system_set_ethtool_settings(struct device *dev, struct device_settings *s)
 	ioctl(sock_ioctl, SIOCETHTOOL, &ifr);
 }
 
+static inline int ethtool_link_mode_test_bit(unsigned int nr, const __u32 *mask)
```
Comment (jow-) [inline system-linux.c:1756]: Please use a `bool` return value for ... well ... boolean returns...

File: system-linux.c [hunk 2]
```diff
@@ -1753,6 +1753,10 @@ system_set_ethtool_settings(struct device *dev, struct device_settings *s)
...
+{
+	return !!(mask[nr / 32] & (1 << (nr % 32)));
```
Comment (jow-) [inline system-linux.c:1758]: ... and subsequently get rid of the `!!(...)`

File: system-linux.c [hunk 3]
```diff
@@ -2362,11 +2366,35 @@ static const struct {
 #endif
 };
 
-static void system_add_link_modes(struct blob_buf *b, __u32 mask)
+static void system_add_link_modes_legacy(struct blob_buf *b, __u32 mask)
 {
-	size_t i;
+	uint32_t i;
```
Comment (jow-) [inline system-linux.c:2371]: The `ARRAY_SIZE(...)` macro call yields a `size_t` type, why the change to `uint32_t`?

File: system-linux.c [hunk 4]
```diff
@@ -2362,11 +2366,35 @@ static const struct {
...
+	{ ETHTOOL_LINK_MODE_1000baseT_Half_BIT, "1000baseT-H" },
+	{ ETHTOOL_LINK_MODE_1000baseT_Full_BIT, "1000baseT-F" },
+	{ ETHTOOL_LINK_MODE_2500baseT_Full_BIT, "2500baseT-F" },
+	{ ETHTOOL_LINK_MODE_5000baseT_Full_BIT, "5000baseT-F" },
+	{ ETHTOOL_LINK_MODE_10000baseT_Full_BIT, "10000baseT-F" },
+};
+
+static void system_add_link_modes(struct blob_buf *b, __u32 *mask)
+{
+	uint32_t i;
```
Comment (jow-) [inline system-linux.c:2395]: Use `size_t`

File: system-linux.c [hunk 5]
```diff
@@ -2588,6 +2616,70 @@ system_if_dump_info(struct device *dev, struct blob_buf *b)
 	return 0;
 }
 
+int
+system_if_dump_info(struct device *dev, struct blob_buf *b)
+{
+	struct {
+		struct ethtool_link_settings req;
+		__u32 link_mode_data[3 * 127];
```
Comment (jow-) [inline system-linux.c:2624]: Where is the magic value `3 * 127` coming from? Why `__u32` instead of `uint32_t`?

File: system-linux.c [hunk 6]
```diff
@@ -2588,6 +2616,70 @@ system_if_dump_info(struct device *dev, struct blob_buf *b)
...
+	} ecmd;
+	struct ifreq ifr = {
+		.ifr_data = (caddr_t)&ecmd,
+	};
+	char *s;
+	void *c;
+	__s8 nwords;
```
Comment (jow-) [inline system-linux.c:2631]: Why `__s8` instead of `int8_t`?

File: system-linux.c [hunk 7]
```diff
@@ -2588,6 +2616,70 @@ system_if_dump_info(struct device *dev, struct blob_buf *b)
...
+	__u32 *supported, *advertising, *lp_advertising;
```
Comment (jow-) [inline system-linux.c:2632]: Why `__u32` instead of `uint32_t`?

File: system-linux.c [hunk 8]
```diff
@@ -2545,8 +2573,8 @@ ethtool_feature_value(const char *ifname, const char *keyname)
 	return active;
 }
 
-int
-system_if_dump_info(struct device *dev, struct blob_buf *b)
+static int
+system_if_dump_info_legacy(struct device *dev, struct blob_buf *b)
```
Comment (jow-) [inline system-linux.c:2577]: If we touch this function declaration anyway we can as well make it `static void`, it can only ever return `0` (which is subsequently ignored by the caller)

File: system-linux.c [hunk 9]
```diff
@@ -2588,6 +2616,70 @@ system_if_dump_info(struct device *dev, struct blob_buf *b)
...
+
+	/* copy link mode bitmaps */
+	nwords = ecmd.req.link_mode_masks_nwords;
+	supported = &ecmd.link_mode_data[0];
+	advertising = &ecmd.link_mode_data[nwords];
+	nwords += ecmd.req.link_mode_masks_nwords;
+	lp_advertising = &ecmd.link_mode_data[nwords];
+
+	c = blobmsg_open_array(b, "link-advertising");
+	system_add_link_modes(b, advertising);
```
Comment (jow-) [inline system-linux.c:2655]: Shouldn't we pass `ecmd.req.link_mode_masks_nwords` to `system_add_link_modes()` and check against it inside there? Currently, none of the modes in `ethtool_link_modes` exceeds the 32 bit boundary, but adding naively adding further ones, like `ETHTOOL_LINK_MODE_25000baseKR_Full_BIT` will lead to out of bound reads on systems only returning one word...

File: system-linux.c [hunk 10]
```diff
@@ -2588,6 +2616,70 @@ system_if_dump_info(struct device *dev, struct blob_buf *b)
...
+
+	c = blobmsg_open_array(b, "link-partner-advertising");
+	system_add_link_modes(b, lp_advertising);
+	blobmsg_close_array(b, c);
+
+	c = blobmsg_open_array(b, "link-supported");
+	system_add_link_modes(b, supported);
+	blobmsg_close_array(b, c);
+
+	s = blobmsg_alloc_string_buffer(b, "speed", 8);
```
Comment (jow-) [inline system-linux.c:2666]: The `speed`, `autoneg` and `hw-tc-offload` attribute handling logic is identical in both the legacy and new code paths as far as I can see, please don't needlessly duplicate that code.

**Review by jow- — CHANGES_REQUESTED** [review]

**Comment by dangowrt** [comment]
Please consider https://github.com/openwrt/netifd/pull/9 instead which completely switches to `ETHTOOL_xLINKSETTINGS` API dropping support for `ETHTOOL_xSET`.

**Comment by dangowrt** [comment]
Closing in favor of #9 which has been merged
